crates/ra_hir/src/code_model.rs

```diff
@@ -849,20 +846,23 @@ impl Trait {
 
     /// Returns an iterator over the whole super trait hierarchy (including the
     /// trait itself).
-    pub fn all_super_traits<'a>(
-        self,
-        db: &'a impl HirDatabase,
-    ) -> impl Iterator<Item = Trait> + 'a {
-        self.all_super_traits_inner(db).unique()
-    }
-
-    fn all_super_traits_inner<'a>(
-        self,
-        db: &'a impl HirDatabase,
-    ) -> impl Iterator<Item = Trait> + 'a {
-        iter::once(self).chain(self.direct_super_traits(db).into_iter().flat_map(move |t| {
-            iter::once(t).chain(Box::new(t.all_super_traits(db)) as Box<dyn Iterator<Item = Trait>>)
-        }))
+    pub fn all_super_traits(self, db: &impl HirDatabase) -> Vec<Trait> {
```

Comment:
it seems like it can return an iterator, though it probably does not makes sense to do.

Reply:
I tried that at first, but that implementation didn't deal correctly with cycles... I think doing it with iterator adapters and handling cycles would be quite complicated, and I'm pretty sure we'd have to keep the set of already visited traits in any case, so it probably wouldn't help :thinking: